TeamCode/src/main/java/org/firstinspires/ftc/teamcode/mm14691/LiftDrive.java

```diff
@@ -65,16 +70,36 @@ public LiftDrive(DcMotorEx motor) {
     }
 
     public ToPosition toDown() {
-        return toPosition(PARAMS.liftDownPosition);
+        return super.toPosition(PARAMS.liftDownPosition);
     }
 
-//    @Override
-//    public void addDebug(@NonNull Telemetry telemetry) {
-//        telemetry.addData(this.getClass().getSimpleName(),
-//                "St: %d, Dwn: %d, Cur: %d, End: %d, Pwr: %f",
-//                getStartTick(), getStartTick() + PARAMS.liftDownPosition,
-//                motor.getCurrentPosition(), getEndTick(),
-//                motor.getPower());
-//    }
+    public void setViperDrive(ViperDrive viperDrive) {
+        this.viperDrive = viperDrive;
+    }
+
+    /**
+     * This runs in the background and updates the viper limits based on the lift arm position.
+     */
+    public class AdjustViperLimits implements Action {
```

Comment:
This is the important bit.  This action runs all the time (just like the motor limit actions) and updates the end limit when the conditions are right.

We'll need to test, but this should play nicely with our existing actions.